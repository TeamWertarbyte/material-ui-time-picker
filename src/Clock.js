import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import { getContrastRatio } from '@material-ui/core/styles/colorManipulator'
import { duration, easing } from '@material-ui/core/styles/transitions'
import classNames from 'classnames'
import { getShortestAngle } from './util'

const styles = (theme) => ({
  root: {
    width: 256,
    height: 256,
    fontFamily: theme.typography.fontFamily,
    cursor: 'default'
  },
  circle: {
    width: 256,
    height: 256,
    borderRadius: '50%',
    background: theme.palette.type === 'light' ? theme.palette.grey[200] : theme.palette.grey[900],
    color: theme.palette.text.primary,
    position: 'relative'
  },
  number: {
    width: 32,
    height: 32,
    left: 'calc(50% - 16px)',
    top: 'calc(50% - 16px)',
    position: 'absolute',
    textAlign: 'center',
    lineHeight: '32px',
    cursor: 'pointer',
    fontSize: '14px',
    pointerEvents: 'none',
    userSelect: 'none',
    '&.selected': {
      color: getContrastRatio(theme.palette.primary.main, theme.palette.common.black) < 7 ? theme.palette.common.white : theme.palette.common.black
    },
    '&.disabled': { opacity: '0.3' }
  },
  smallNumber: {
    fontSize: '12px',
    color: theme.palette.text.secondary
  },
  pointer: {
    width: 'calc(50% - 20px)',
    height: 2,
    backgroundColor: theme.palette.primary.main,
    position: 'absolute',
    left: '50%',
    top: 'calc(50% - 1px)',
    transformOrigin: 'left center',
    pointerEvents: 'none'
  },
  animatedPointer: {
    transition: `all ${duration.short}ms ${easing.easeInOut}`
  },
  smallPointer: {
    width: 'calc(50% - 52px)'
  },
  innerDot: {
    backgroundColor: theme.palette.primary.main,
    position: 'absolute',
    top: -4 + 1,
    left: -4,
    width: 8,
    height: 8,
    borderRadius: '50%'
  },
  outerDot: {
    border: `16px solid ${theme.palette.primary.main}`,
    borderWidth: 16,
    position: 'absolute',
    top: -16 + 1,
    right: -16,
    width: 0,
    height: 0,
    borderRadius: '50%',
    boxSizing: 'content-box'
  },
  outerDotOdd: {
    background: getContrastRatio(theme.palette.primary.main, theme.palette.common.black) < 7 ? theme.palette.common.white : theme.palette.common.black,
    width: 4,
    height: 4,
    borderWidth: 14
  }
})

const size = 256

class Clock extends React.PureComponent {
  constructor (props) {
    super(props)
    this.state = { touching: false, angle: getPointerAngle(props.value, props.mode) }
  }

  componentWillReceiveProps ({ value, mode }) {
    if (mode !== this.props.mode || value !== this.props.value) {
      this.setState({ angle: getShortestAngle(this.state.angle, getPointerAngle(value, mode)) })
    }
  }

  disableAnimatedPointer = () => this.setState({ touching: true })
  enableAnimatedPointer = () => this.setState({ touching: false })

  handleTouchMove = (e) => {
    e.preventDefault() // prevent scrolling behind the clock on iOS
    const rect = e.target.getBoundingClientRect()
    this.movePointer(e.changedTouches[0].clientX - rect.left, e.changedTouches[0].clientY - rect.top)
  }

  handleTouchEnd = (e) => {
    this.handleTouchMove(e)
    this.enableAnimatedPointer()
  }

  handleMouseMove = (e) => {
    // MouseEvent.which is deprecated, but MouseEvent.buttons is not supported in Safari
    if (e.buttons === 1 || e.which === 1) {
      const rect = e.target.getBoundingClientRect()
      this.movePointer(e.clientX - rect.left, e.clientY - rect.top)
    }
  }

  handleClick = (e) => {
    const rect = e.target.getBoundingClientRect()
    this.movePointer(e.clientX - rect.left, e.clientY - rect.top)
  }

  movePointer (x, y) {
    const value = getPointerValue(x, y, this.props.mode)
    if (value !== this.props.value && this.props.onChange != null) {
      this.props.onChange(value)
    }
  }

  render () {
    const { classes, mode, value, minutesStep, ...other } = this.props
    const { touching } = this.state

    return (
      <div className={classes.root} {...other}>
        <div
          className={classes.circle}
          onTouchMove={this.handleTouchMove}
          onMouseMove={this.handleMouseMove}
          onTouchStart={this.disableAnimatedPointer}
          onMouseDown={this.disableAnimatedPointer}
          onTouchEnd={this.handleTouchEnd}
          onMouseUp={this.enableAnimatedPointer}
          onClick={this.handleClick}
        >
          <div className={classNames(classes.pointer, { [classes.smallPointer]: mode === '24h' && (value === 0 || value > 12), [classes.animatedPointer]: !touching })} style={{
            transform: `rotate(${this.state.angle}deg)`
          }}>
            <div className={classes.innerDot} />
            <div className={classNames(classes.outerDot, { [classes.outerDotOdd]: mode === 'minutes' && value % 5 !== 0 })} />
          </div>
          {mode === '12h' && getNumbers(12, { size }).map((digit, i) => (
            <span
              key={digit.display}
              className={classNames(classes.number, { selected: value === digit.display })}
              style={{
                transform: `translate(${digit.translateX}px, ${digit.translateY}px)`
              }}
            >
              {digit.display}
            </span>
          ))}
          {mode === '24h' && getNumbers(12, { size }).map((digit, i) => (
            <span
              key={digit.display}
              className={classNames(classes.number, { selected: value === digit.display })}
              style={{
                transform: `translate(${digit.translateX}px, ${digit.translateY}px)`
              }}
            >
              {digit.display}
            </span>
          ))}
          {mode === '24h' && getNumbers(12, { size: size - 64, start: 13 }).map((digit, i) => (
            <span
              key={digit.display}
              className={classNames(classes.number, classes.smallNumber, { selected: value === digit.display || (digit.display === 24 && value === 0) })}
              style={{
                transform: `translate(${digit.translateX}px, ${digit.translateY}px)`
              }}
            >
              {digit.display === 24 ? '00' : digit.display}
            </span>
          ))}
          {mode === 'minutes' && getNumbers(12, { size, start: 5, step: 5 }).map((digit, i) => (
            <span
              key={digit.display}
              className={classNames(
                classes.number,
                { selected: value === digit.display || (digit.display === 60 && value === 0), disabled: digit.display % (this.props.minutesStep) })}
              style={{
                transform: `translate(${digit.translateX}px, ${digit.translateY}px)`
              }}
            >
              {digit.display === 60 ? '00' : digit.display}
            </span>
          ))}
        </div>
      </div>
    )
  }
}

Clock.propTypes = {
  /** Steps between minutes. */
  minutesStep: PropTypes.number,
  /** Sets the mode of this clock. It can either select hours (supports 12- and 24-hour-clock) or minutes. */
  mode: PropTypes.oneOf(['12h', '24h', 'minutes']).isRequired,
  /** Callback that is called with the new hours/minutes (as a number) when the value is changed. */
  onChange: PropTypes.func,
  /** The value of the clock. */
  value: PropTypes.number.isRequired
}

Clock.defaultProps = {
  minutesStep: 1
}

export default withStyles(styles)(Clock)

function getNumbers (count, { size, start = 1, step = 1 }) {
  return Array.apply(null, Array(count)).map((_, i) => ({
    display: i * step + start,
    translateX: (size / 2 - 20) * Math.cos(2 * Math.PI * (i - 2) / count),
    translateY: (size / 2 - 20) * Math.sin(2 * Math.PI * (i - 2) / count)
  }))
}

function getPointerAngle (value, mode) {
  switch (mode) {
    case '12h':
      return 360 / 12 * (value - 3)
    case '24h':
      return 360 / 12 * (value % 12 - 3)
    case 'minutes':
      return 360 / 60 * (value - 15)
  }
}

function getPointerValue (x, y, mode) {
  let angle = Math.atan2(size / 2 - x, size / 2 - y) / Math.PI * 180
  if (angle < 0) {
    angle = 360 + angle
  }

  switch (mode) {
    case '12h': {
      const value = 12 - Math.round(angle * 12 / 360)
      return value === 0 ? 12 : value
    }
    case '24h': {
      const radius = Math.sqrt(Math.pow(size / 2 - x, 2) + Math.pow(size / 2 - y, 2))
      let value = 12 - Math.round(angle * 12 / 360)
      if (value === 0) {
        value = 12
      }
      if (radius < size / 2 - 32) {
        value = value === 12 ? 0 : value + 12
      }
      return value
    }
    case 'minutes': {
      const value = Math.round(60 - 60 * angle / 360)
      return value === 60 ? 0 : value
    }
  }
}
