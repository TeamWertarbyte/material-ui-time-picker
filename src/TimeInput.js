import React from 'react'
import PropTypes from 'prop-types'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import Button from '@material-ui/core/Button'
import Input from '@material-ui/core/Input'
import { withStyles } from '@material-ui/core/styles'
import TimePicker from './TimePicker'
import { formatHours, twoDigits } from './util'

const styles = {
  header: {
    borderTopLeftRadius: 2,
    borderTopRightRadius: 2
  },
  body: {
    paddingBottom: 20
  }
}

class TimeInput extends React.Component {
  constructor (props) {
    super(props)
    const defaultValue = new Date()
    defaultValue.setSeconds(0)
    defaultValue.setMilliseconds(0)
    const initValue = props.value || props.defaultValue || defaultValue
    this.state = { open: false, initValue, value: initValue }
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.value !== this.props.value) {
      this.setState({ initValue: nextProps.value, value: nextProps.value })
    }
  }

  showDialog = () => this.setState({ open: true })

  handleChange = (value) => {
    if (this.props.updateImmediately && this.props.onChange != null) {
      this.props.onChange(value)
    }
    this.setState({ value })
  }

  handleOk = () => {
    if (!this.props.updateImmediately && this.props.onChange != null) {
      this.props.onChange(this.state.value)
    }
    this.setState({ open: false, initValue: this.state.value })
  }

  handleCancel = () => {
    if (this.props.updateImmediately && this.props.onChange != null) {
      this.props.onChange(this.state.initValue)
    }
    this.setState({ open: false, value: this.state.initValue })
  }

  render () {
    const {
      autoOk,
      cancelLabel,
      classes,
      defaultValue, // eslint-disable-line
      disabled: disabledProp,
      mode,
      okLabel,
      onChange, // eslint-disable-line
      value: valueProp, // eslint-disable-line
      minutesStep,
      selectOnClose,
      updateImmediately,
      ...other
    } = this.props

    const { initValue, value } = this.state
    const inputValue = updateImmediately ? value : initValue

    const { hours, isPm } = formatHours(inputValue.getHours(), mode)
    const formattedValue = mode === '12h'
      ? `${hours}:${twoDigits(inputValue.getMinutes())} ${isPm ? 'pm' : 'am'}`
      : `${twoDigits(inputValue.getHours())}:${twoDigits(inputValue.getMinutes())}`

    const { muiFormControl } = this.context
    const disabled = disabledProp || (muiFormControl != null && muiFormControl.disabled)

    return [
      <Input
        {...other}
        disabled={disabled}
        onClick={!disabled ? this.showDialog : null}
        value={formattedValue}
        readOnly
        key='TimeInput-input'
        classes={{ input: classes.input }}
      />,
      <Dialog
        maxWidth='xs'
        open={this.state.open}
        key='TimeInput-dialog'
        onClose={selectOnClose ? this.handleOk : this.handleCancel}
      >
        <TimePicker
          mode={mode}
          minutesStep={minutesStep}
          value={value}
          onChange={this.handleChange}
          onMinutesSelected={autoOk ? this.handleOk : null}
          classes={{ header: classes.header, body: classes.body }}
        />
        <DialogActions>
          <Button onClick={this.handleCancel} color='primary'>{cancelLabel}</Button>
          <Button onClick={this.handleOk} color='primary'>{okLabel}</Button>
        </DialogActions>
      </Dialog>
    ]
  }
}

TimeInput.propTypes = {
  /** If true, automatically accept and close the picker on set minutes. */
  autoOk: PropTypes.bool,
  /** Override the label of the cancel button. */
  cancelLabel: PropTypes.string,
  /** The initial value of the time picker. */
  defaultValue: PropTypes.instanceOf(Date),
  /** Steps between minutes. */
  minutesStep: PropTypes.number,
  /** Sets the clock mode, 12-hour or 24-hour clocks are supported. */
  mode: PropTypes.oneOf(['12h', '24h']),
  /** Override the label of the ok button. */
  okLabel: PropTypes.string,
  /** Callback that is called with the new date (as Date instance) when the value is changed. */
  onChange: PropTypes.func,
  /** Sets the date when dialog is closed (clicking background). */
  selectOnClose: PropTypes.bool,
  /** Updates the input field when clock hands are moved */
  updateImmediately: PropTypes.bool,
  /** The value of the time picker, for use in controlled mode. */
  value: PropTypes.instanceOf(Date)
}

TimeInput.defaultProps = {
  autoOk: false,
  cancelLabel: 'Cancel',
  minutesStep: 1,
  mode: '12h',
  okLabel: 'Ok',
  selectOnClose: false,
  updateImmediately: false
}

TimeInput.contextTypes = {
  muiFormControl: PropTypes.object
}

export default withStyles(styles)(TimeInput)
