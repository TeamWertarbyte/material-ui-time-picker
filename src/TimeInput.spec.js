/* eslint-env jest */
import React from 'react'
import { mount } from 'enzyme'
import MockDate from 'mockdate'
import { unwrap } from '@material-ui/core/test-utils'
import Button from '@material-ui/core/Button'
import Input from '@material-ui/core/Input'
import TimeInput from './TimeInput'
import Clock from './Clock'
import * as testUtils from '../test/utils'

describe('<TimeInput />', () => {
  describe('24h', () => {
    it('matches the snapshot', () => {
      const originalGetHours = Date.prototype.getHours
      Date.prototype.getHours = Date.prototype.getUTCHours // eslint-disable-line
      const tree = mount(<TimeInput defaultValue={new Date('2017-10-15T13:37Z')} mode='24h' />)
      expect(tree).toMatchSnapshot()
      Date.prototype.getHours = originalGetHours // eslint-disable-line
    })

    it('displays the formatted time', () => {
      const tree = mount(<TimeInput value={new Date(2017, 10, 15, 13, 37, 0, 0)} mode='24h' />)
      expect(getValue(tree)).toBe('13:37')

      const tree2 = mount(<TimeInput value={new Date(2017, 10, 15, 1, 23, 0, 0)} mode='24h' />)
      expect(getValue(tree2)).toBe('01:23')
    })

    it('disables the input if the disabled prop is set to true', () => {
      const tree = mount(<TimeInput value={new Date(2017, 10, 15, 13, 37, 0, 0)} mode='24h' disabled />)
      expect(tree.find(Input).prop('disabled')).toBe(true)
    })
  })

  describe('12h', () => {
    it('matches the snapshot', () => {
      const originalGetHours = Date.prototype.getHours
      Date.prototype.getHours = Date.prototype.getUTCHours // eslint-disable-line
      const tree = mount(<TimeInput defaultValue={new Date('2017-10-15T13:37Z')} mode='12h' />)
      expect(tree).toMatchSnapshot()
      Date.prototype.getHours = originalGetHours // eslint-disable-line
    })

    it('displays the formatted time', () => {
      const tree = mount(<TimeInput value={new Date(2017, 10, 15, 13, 37, 0, 0)} mode='12h' />)
      expect(getValue(tree)).toBe('1:37 pm')

      const tree2 = mount(<TimeInput value={new Date(2017, 10, 15, 1, 23, 0, 0)} mode='12h' />)
      expect(getValue(tree2)).toBe('1:23 am')

      const tree3 = mount(<TimeInput value={new Date(2017, 10, 15, 0, 0, 0, 0)} mode='12h' />)
      expect(getValue(tree3)).toBe('12:00 am')

      const tree4 = mount(<TimeInput value={new Date(2017, 10, 15, 0, 0, 0, 0)} mode='12h' />)
      expect(getValue(tree4)).toBe('12:00 am')
    })
  })

  it('supports controlled mode', () => {
    const tree = mount(<TimeInput value={new Date(2017, 10, 15, 13, 37, 0, 0)} mode='24h' />)
    expect(getValue(tree)).toBe('13:37')

    tree.setProps({ value: new Date(2017, 10, 15, 14, 42, 0, 0) })
    expect(getValue(tree)).toBe('14:42')
  })

  describe('uncontrolled mode', () => {
    afterEach(() => {
      MockDate.reset()
    })

    it('supports uncontrolled mode with a default value', () => {
      const tree = mount(<TimeInput defaultValue={new Date(2017, 10, 15, 13, 37, 0, 0)} mode='24h' />)
      expect(getValue(tree)).toBe('13:37')
    })

    it('uses the current time if no value or default value is set', () => {
      MockDate.set(new Date(2017, 10, 15, 13, 37, 0, 0))
      const tree = mount(<TimeInput mode='24h' />)
      expect(getValue(tree)).toBe('13:37')
    })
  })

  describe('TimePicker dialog', () => {
    it('opens when clicking the input', () => {
      const UnstyledTimeInput = unwrap(TimeInput)
      const tree = mount(<UnstyledTimeInput classes={{}} />)
      tree.simulate('click')
      expect(tree.state('open')).toBe(true)
    })

    it('opens when clicking the input if it is disabled', () => {
      const UnstyledTimeInput = unwrap(TimeInput)
      const tree = mount(<UnstyledTimeInput classes={{}} disabled />)
      tree.simulate('click')
      expect(tree.state('open')).toBe(false)
    })

    it('closes when clicking ok', () => {
      const UnstyledTimeInput = unwrap(TimeInput)
      const tree = mount(<UnstyledTimeInput classes={{}} />)
      tree.simulate('click')
      tree.find(Button).at(1).simulate('click')
      expect(tree.state('open')).toBe(false)
    })

    it('updates uses the new time when clicking ok', () => {
      const changeHandler = jest.fn()
      MockDate.set(new Date(2017, 10, 15, 13, 37, 0, 0))
      const tree = mount(<TimeInput classes={{}} mode='24h' onChange={changeHandler} />)

      tree.simulate('click')
      getCircle(tree.find(Clock))
        .simulate('click', testUtils.stubClickEvent(128, 30)) // click on 12
        .simulate('mouseup', testUtils.stubClickEvent(128, 30)) // click on 12
      tree.find(Button).at(1).simulate('click')

      expect(getValue(tree)).toBe('12:37')
      expect(changeHandler).toHaveBeenCalled()
      const time = changeHandler.mock.calls[0][0]
      expect(time.getHours()).toBe(12)
      expect(time.getMinutes()).toBe(37)
    })

    it('closes when clicking cancel', () => {
      const tree = mount(<TimeInput />)
      tree.simulate('click')
      tree.find(Button).at(0).simulate('click')
    })

    it('discards the new time when clicking cancel', () => {
      const changeHandler = jest.fn()
      MockDate.set(new Date(2017, 10, 15, 13, 37, 0, 0))
      const tree = mount(<TimeInput classes={{}} mode='24h' onChange={changeHandler} />)

      tree.simulate('click')
      getCircle(tree.find(Clock))
        .simulate('click', testUtils.stubClickEvent(128, 30)) // click on 12
        .simulate('mouseup', testUtils.stubClickEvent(128, 30)) // click on 12
      tree.find(Button).at(0).simulate('click')

      expect(getValue(tree)).toBe('13:37') // unchanged
    })
  })
})

function getValue (timeInput) {
  return timeInput.find('input[type="text"]').getDOMNode().value
}

function getCircle (clock) {
  return clock.childAt(0).children('div').childAt(0)
}
