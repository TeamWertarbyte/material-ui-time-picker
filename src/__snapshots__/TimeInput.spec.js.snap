// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeInput /> 12h matches the snapshot 1`] = `
<WithStyles(TimeInput)
  defaultValue={2017-10-15T13:37:00.000Z}
  mode="12h"
>
  <TimeInput
    autoOk={false}
    cancelLabel="Cancel"
    classes={
      Object {
        "body": "TimeInput-body-2",
        "header": "TimeInput-header-1",
      }
    }
    defaultValue={2017-10-15T13:37:00.000Z}
    minutesStep={1}
    mode="12h"
    okLabel="Ok"
    selectOnClose={false}
    updateImmediately={false}
  >
    <WithStyles(Input)
      classes={
        Object {
          "input": undefined,
        }
      }
      disabled={false}
      key="TimeInput-input"
      onClick={[Function]}
      readOnly={true}
      value="1:37 pm"
    >
      <Input
        classes={
          Object {
            "disabled": "MuiInput-disabled-6",
            "error": "MuiInput-error-8",
            "focused": "MuiInput-focused-5",
            "formControl": "MuiInput-formControl-4",
            "fullWidth": "MuiInput-fullWidth-10",
            "input": "MuiInput-input-11",
            "inputMarginDense": "MuiInput-inputMarginDense-12",
            "inputMultiline": "MuiInput-inputMultiline-13",
            "inputType": "MuiInput-inputType-14",
            "inputTypeSearch": "MuiInput-inputTypeSearch-15",
            "multiline": "MuiInput-multiline-9",
            "root": "MuiInput-root-3",
            "underline": "MuiInput-underline-7",
          }
        }
        disableUnderline={false}
        disabled={false}
        fullWidth={false}
        multiline={false}
        onClick={[Function]}
        readOnly={true}
        type="text"
        value="1:37 pm"
      >
        <div
          className="MuiInput-root-3 MuiInput-underline-7"
          onClick={[Function]}
        >
          <input
            className="MuiInput-input-11"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            readOnly={true}
            type="text"
            value="1:37 pm"
          />
        </div>
      </Input>
    </WithStyles(Input)>
    <WithStyles(Dialog)
      key="TimeInput-dialog"
      maxWidth="xs"
      onClose={[Function]}
      open={false}
    >
      <Dialog
        TransitionComponent={[Function]}
        classes={
          Object {
            "paper": "MuiDialog-paper-17",
            "paperFullScreen": "MuiDialog-paperFullScreen-22",
            "paperFullWidth": "MuiDialog-paperFullWidth-21",
            "paperWidthMd": "MuiDialog-paperWidthMd-20",
            "paperWidthSm": "MuiDialog-paperWidthSm-19",
            "paperWidthXs": "MuiDialog-paperWidthXs-18",
            "root": "MuiDialog-root-16",
          }
        }
        disableBackdropClick={false}
        disableEscapeKeyDown={false}
        fullScreen={false}
        fullWidth={false}
        maxWidth="xs"
        onClose={[Function]}
        open={false}
        transitionDuration={
          Object {
            "enter": 225,
            "exit": 195,
          }
        }
      >
        <WithStyles(Modal)
          BackdropProps={
            Object {
              "transitionDuration": Object {
                "enter": 225,
                "exit": 195,
              },
            }
          }
          className="MuiDialog-root-16"
          disableBackdropClick={false}
          disableEscapeKeyDown={false}
          onClose={[Function]}
          open={false}
          role="dialog"
        >
          <Modal
            BackdropComponent={[Function]}
            BackdropProps={
              Object {
                "transitionDuration": Object {
                  "enter": 225,
                  "exit": 195,
                },
              }
            }
            className="MuiDialog-root-16"
            classes={
              Object {
                "hidden": "MuiModal-hidden-24",
                "root": "MuiModal-root-23",
              }
            }
            disableAutoFocus={false}
            disableBackdropClick={false}
            disableEnforceFocus={false}
            disableEscapeKeyDown={false}
            disableRestoreFocus={false}
            hideBackdrop={false}
            keepMounted={false}
            manager={
              ModalManager {
                "containers": Array [],
                "data": Array [],
                "handleContainerOverflow": true,
                "hideSiblingNodes": true,
                "modals": Array [],
              }
            }
            onClose={[Function]}
            open={false}
            role="dialog"
          />
        </WithStyles(Modal)>
      </Dialog>
    </WithStyles(Dialog)>
  </TimeInput>
</WithStyles(TimeInput)>
`;

exports[`<TimeInput /> 24h matches the snapshot 1`] = `
<WithStyles(TimeInput)
  defaultValue={2017-10-15T13:37:00.000Z}
  mode="24h"
>
  <TimeInput
    autoOk={false}
    cancelLabel="Cancel"
    classes={
      Object {
        "body": "TimeInput-body-2",
        "header": "TimeInput-header-1",
      }
    }
    defaultValue={2017-10-15T13:37:00.000Z}
    minutesStep={1}
    mode="24h"
    okLabel="Ok"
    selectOnClose={false}
    updateImmediately={false}
  >
    <WithStyles(Input)
      classes={
        Object {
          "input": undefined,
        }
      }
      disabled={false}
      key="TimeInput-input"
      onClick={[Function]}
      readOnly={true}
      value="13:37"
    >
      <Input
        classes={
          Object {
            "disabled": "MuiInput-disabled-6",
            "error": "MuiInput-error-8",
            "focused": "MuiInput-focused-5",
            "formControl": "MuiInput-formControl-4",
            "fullWidth": "MuiInput-fullWidth-10",
            "input": "MuiInput-input-11",
            "inputMarginDense": "MuiInput-inputMarginDense-12",
            "inputMultiline": "MuiInput-inputMultiline-13",
            "inputType": "MuiInput-inputType-14",
            "inputTypeSearch": "MuiInput-inputTypeSearch-15",
            "multiline": "MuiInput-multiline-9",
            "root": "MuiInput-root-3",
            "underline": "MuiInput-underline-7",
          }
        }
        disableUnderline={false}
        disabled={false}
        fullWidth={false}
        multiline={false}
        onClick={[Function]}
        readOnly={true}
        type="text"
        value="13:37"
      >
        <div
          className="MuiInput-root-3 MuiInput-underline-7"
          onClick={[Function]}
        >
          <input
            className="MuiInput-input-11"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            readOnly={true}
            type="text"
            value="13:37"
          />
        </div>
      </Input>
    </WithStyles(Input)>
    <WithStyles(Dialog)
      key="TimeInput-dialog"
      maxWidth="xs"
      onClose={[Function]}
      open={false}
    >
      <Dialog
        TransitionComponent={[Function]}
        classes={
          Object {
            "paper": "MuiDialog-paper-17",
            "paperFullScreen": "MuiDialog-paperFullScreen-22",
            "paperFullWidth": "MuiDialog-paperFullWidth-21",
            "paperWidthMd": "MuiDialog-paperWidthMd-20",
            "paperWidthSm": "MuiDialog-paperWidthSm-19",
            "paperWidthXs": "MuiDialog-paperWidthXs-18",
            "root": "MuiDialog-root-16",
          }
        }
        disableBackdropClick={false}
        disableEscapeKeyDown={false}
        fullScreen={false}
        fullWidth={false}
        maxWidth="xs"
        onClose={[Function]}
        open={false}
        transitionDuration={
          Object {
            "enter": 225,
            "exit": 195,
          }
        }
      >
        <WithStyles(Modal)
          BackdropProps={
            Object {
              "transitionDuration": Object {
                "enter": 225,
                "exit": 195,
              },
            }
          }
          className="MuiDialog-root-16"
          disableBackdropClick={false}
          disableEscapeKeyDown={false}
          onClose={[Function]}
          open={false}
          role="dialog"
        >
          <Modal
            BackdropComponent={[Function]}
            BackdropProps={
              Object {
                "transitionDuration": Object {
                  "enter": 225,
                  "exit": 195,
                },
              }
            }
            className="MuiDialog-root-16"
            classes={
              Object {
                "hidden": "MuiModal-hidden-24",
                "root": "MuiModal-root-23",
              }
            }
            disableAutoFocus={false}
            disableBackdropClick={false}
            disableEnforceFocus={false}
            disableEscapeKeyDown={false}
            disableRestoreFocus={false}
            hideBackdrop={false}
            keepMounted={false}
            manager={
              ModalManager {
                "containers": Array [],
                "data": Array [],
                "handleContainerOverflow": true,
                "hideSiblingNodes": true,
                "modals": Array [],
              }
            }
            onClose={[Function]}
            open={false}
            role="dialog"
          />
        </WithStyles(Modal)>
      </Dialog>
    </WithStyles(Dialog)>
  </TimeInput>
</WithStyles(TimeInput)>
`;
